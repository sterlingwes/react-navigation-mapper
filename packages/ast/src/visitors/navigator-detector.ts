import {
  isCallExpression,
  isIdentifier,
  isImportSpecifier,
  isJsxAttribute,
  isJsxExpression,
  isJsxSelfClosingElement,
  isStringLiteral,
} from "typescript";

import {
  resolveNavigationStackImportFactoryName,
  resolveNavigatorNamespaceFromFactoryCall,
} from "./utils/navigators";
import { jsxElementIsScreen } from "./utils/stack-screen";
import { ComponentAwareVisitor } from "./mixable/component-aware-visitor";

type Screen = {
  name: string;
  component: string;
};

type Stack = {
  name: string;
  namespace: string;
  screens: Array<Screen>;
};

interface GlobalState {
  stacks: Record<string, Stack>;
}

interface ModuleState {
  stackFactory: string;
  stackNamespace: string;
  stackComponentName: string;
}

// check if import specifier exists
// if so, check if factory is called and track identifier
// if so, check for identifier in JSX tags
// if so, capture params passed to those components as screen routes
//
// maybe this query lib can help? https://github.com/phenomnomnominal/tsquery
// can filter to files that have the import specifier then query further

export class NavigatorDetectorVisitor extends ComponentAwareVisitor<
  GlobalState,
  ModuleState
> {
  constructor() {
    super({
      initialGlobalState: {
        stacks: {},
      },
      initialModuleState: {
        stackFactory: "",
        stackNamespace: "",
        stackComponentName: "",
      },
    });

    this.addVisitorCases([
      this.case(isImportSpecifier, (node, { moduleState }) => {
        const factoryMatch = resolveNavigationStackImportFactoryName(node);
        if (factoryMatch) {
          moduleState.stackFactory = factoryMatch;
        }
      }),

      this.case(isCallExpression, (node, { moduleState }) => {
        const namespaceMatch = resolveNavigatorNamespaceFromFactoryCall(
          node,
          moduleState.stackFactory
        );
        if (namespaceMatch) {
          moduleState.stackNamespace = namespaceMatch;
        }
      }),

      // TODO: handle destructuring of Screen instead of property access on stack identifier
      this.case(isJsxSelfClosingElement, (node, { moduleState }) => {
        if (jsxElementIsScreen(node, moduleState.stackNamespace)) {
          const screen: Screen = { name: "", component: "" };
          node.attributes.properties.forEach((attr) => {
            if (!isJsxAttribute(attr) || !isIdentifier(attr.name)) return;

            if (
              attr.name.escapedText === "name" &&
              attr.initializer &&
              // TODO: handle other cases besides string literal
              isStringLiteral(attr.initializer)
            ) {
              screen.name = attr.initializer.text;
            }

            if (
              attr.name.escapedText === "component" &&
              attr.initializer &&
              isJsxExpression(attr.initializer) &&
              attr.initializer.expression &&
              isIdentifier(attr.initializer.expression)
            ) {
              screen.component = attr.initializer.expression
                .escapedText as string;
            }
          });

          if (screen.name && screen.component) {
            this.addScreen(screen);
          }
        }
      }),
    ]);
  }

  private addScreen(screen: Screen) {
    const component = this.ensureComponent(`when adding screen ${screen.name}`);

    if (!this.globalState.stacks[component]) {
      this.initializeStack(component);
    }

    const { stacks } = this.globalState;
    stacks[component] = {
      ...stacks[component],
      screens: [...stacks[component].screens, screen],
    };
  }

  private initializeStack(namespace: string) {
    const component = this.ensureComponent(
      `when initializing stack ${namespace}`
    );

    if (!this.globalState.stacks[namespace]) {
      this.globalState.stacks[component] = {
        name: component,
        namespace,
        screens: [],
      };
    } else {
      console.warn(
        `Stack ${namespace} already exists, visitor in unexpected state`
      );
    }
  }

  private ensureComponent(detail = "") {
    const component = this.moduleState.currentComponent;
    if (!component) {
      throw new Error(`No component name set, expected to find one: ${detail}`);
    }

    return component;
  }
}
