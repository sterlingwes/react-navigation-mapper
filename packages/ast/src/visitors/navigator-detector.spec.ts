import fs from "fs";
import { describe, it, expect, beforeEach } from "bun:test";
import ts from "typescript";
import { NavigatorDetectorVisitor } from "./navigator-detector";

describe("NavigatorDetector visitor", () => {
  describe("single navigator stack with multiple screens", () => {
    let file: ts.SourceFile;

    beforeEach(() => {
      const fileName =
        "submodules/react-navigation/example/src/Screens/NativeStack.tsx";
      file = ts.createSourceFile(
        fileName,
        fs.readFileSync(fileName).toString(),
        ts.ScriptTarget.ES2015,
        /*setParentNodes */ true
      );
    });

    it("should detect all screens and save references", () => {
      const visitor = new NavigatorDetectorVisitor();
      visitor.visit(file);

      expect(visitor.globalState.stacks).toMatchSnapshot();
    });
  });
});
