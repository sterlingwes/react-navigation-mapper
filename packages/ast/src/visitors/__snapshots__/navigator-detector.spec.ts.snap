// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigatorDetector visitor single navigator stack with multiple screens should detect all screens and save references 1`] = `
{
  "NativeStackScreen": {
    "name": "NativeStackScreen",
    "namespace": "NativeStackScreen",
    "screens": [
      {
        "component": "ArticleScreen",
        "importPath": undefined,
        "name": "Article",
        "options": {
          "headerLargeTitle": true,
          "headerLargeTitleShadowVisible": false,
          "title": null,
        },
      },
      {
        "component": "NewsFeedScreen",
        "importPath": undefined,
        "name": "NewsFeed",
        "options": {
          "fullScreenGestureEnabled": true,
          "title": "Feed",
        },
      },
      {
        "component": "AlbumsScreen",
        "importPath": undefined,
        "name": "Albums",
        "options": {
          "headerBlurEffect": "light",
          "headerTransparent": true,
          "presentation": "modal",
          "title": "Albums",
        },
      },
    ],
    "source": "submodules/react-navigation/example/src/Screens/NativeStack.tsx",
  },
}
`;
